### ISSUE TITLE
Add support for 3D buildings in MapboxGL

### ISSUE THREAD

**Issue by stirringhalo** [osm2vectortiles/osm2vectortiles#454]
In light of added support for fill-extrude-height and fill-extrude-base, I'd like to add the height attributes to OSM2Vectortiles backend so at a later date, the tm2source and style (probably Liberty)  can pick it up.

I'll take this on. Steps I can see:

- [ ] adjust imposm mapping to add it in
- [ ] confirm that a demo tm2source can pick it up (solely will render buildings into the mbtiles)
- [ ] test display in a viewer

**Comment by lukasmartinelli** [comment]
Cool. In the v3 we use the following approach of Paul Norman ([OSM Clear](https://github.com/ClearTables/osm-clear/blob/master/osm-clear.tm2source/data.yml) to calculate `render_height`:

```sql
-- 3.66 is obtained by correlating the heights and level data in OSM
-- raising to a power < 1 reduces the height range displayed which looks better
least(greatest(3, COALESCE(height,levels*3.66,5)),400)^.7 AS render_height
```

**Comment by kekscom** [comment]
Hi,
I'm considering to join your efforts and go for Vector Tile support in OSM Buildings.
Is there a chance to include a lot more attributes, i.e. building levels, roof shapes etc.?
Packing as JSON string would be fine.
With that, I could also imagine an extended version of MapboxGL in terms of 3D buildings.

**Comment by stirringhalo** [comment]
@kekscom 
Yep, I could definitely include more attributes, though they won't be rendered as complex rooftops!

With regards to 3D and MapboxGL, I'm not sure it's 100% at the point where it replaces something like Cesium in capability. I have a working offline demo that uses OSM2World and 3DCityDB to fill a database with buildings and dump them to tiles to view in Cesium. It's a powerful solution for handling lots of buildings with the full capabilities of drawing complex rooftops as opposed to basic extrusions. Only hurdle is hardware needs to view lots of buildings in a downtown area when in KML formats.

**Comment by stirringhalo** [comment]
Okay, I did the first demos last night. I got it to draw in 3D, but I don't have any proper height dependent 3D drawing yet. Might be an issue that I did Switzerland, mostly centered over Zurich so not much in the way of complex buildings. I will try New York next. Unfortunately the fill-extrude-height and fill-extrude-base requires "stops", adding a layer of complexity to my demo style.

Second issue that I may face, what "order" do I do in the tm2source for buildings? Right now, it's `ORDER BY ST_YMin(ST_Envelope(geometry)) DESC`. But thinking about it now, I may need to change this to be height dependent, drawing the lowest layers first, potentially sorting on min_level/min_height. Hmmm.

**Comment by kekscom** [comment]
@stirringhalo First I needed to know whether you would be interested in adding more information. As coverage with extended buildings tags varies a lot, it might not make sense to predefine e fixed attribute set.

Once there is an agreement about that, I'm fine to work on the renderer.

**Comment by stirringhalo** [comment]
@kekscom I have no problem adding more tag information.

**Comment by stirringhalo** [comment]
![image](https://cloud.githubusercontent.com/assets/14035518/19604253/754bb1a4-9782-11e6-9438-aa5a68bd2732.png)

Have to figure out why building:part is missing, I'm not as familiar with imposm3 mapping as I am imposm2. 

I'd like to put forward a tm2source eventually with all the tags, but it will explicitly not be compatible with Streets as much as reasonably possible.

**Comment by stirringhalo** [comment]
I figured out the bug, it was a missing set of keys in mapping.yml!

Looking great so far, I'm adding the additional features and I'm testing data-driven colouring using the colour functionality. I would love for Mapbox to add support for fallback values in the "identity" function when a value does not exist. For now, I'm just fill with default values in DB but it's definitely ugly.

**Comment by stirringhalo** [comment]
More progress, I have building colours working. Next up will be materials for buildings.

![screenshot](https://cloud.githubusercontent.com/assets/14035518/19610085/91542040-97a8-11e6-817d-176e8408273b.png)

If we could get rooftop drawing done, that would be absolutely amazing!

**Comment by lukasmartinelli** [comment]
> Have to figure out why building:part is missing, I'm not as familiar with imposm3 mapping as I am imposm2. Any thoughts @lukasmartinelli ?

Just note here which OSM source data you require for the 3D buildings.

> Decided against rendering the materials as that would get extremely complicated and may not look very good with the limited pattern fill capabilities

Wow looks very nice ❤️

**Comment by stirringhalo** [comment]
@lukasmartinelli
For all, see the enhancement/add_height branch. It's got all the current progress for this.

I'm waiting on https://github.com/mapbox/mapbox-gl-style-spec/issues/480 to see how default fall-back styles will be handled for the identity function.

Feel free to rip these out and put them into the open tm2source and associated import workflow.

**Comment by stirringhalo** [comment]
Integrated with OpenMapTiles....

https://openmaptiles.github.io/klokantech-3d-gl-style/#14.47/40.7086/-74.0049/16/56

**Comment by nicooprat** [comment]
Amazing! Any idea what's happening with the Eiffel Tower though? https://openmaptiles.github.io/klokantech-3d-gl-style/#15.85/48.8593/2.2959/0/57

**Comment by stirringhalo** [comment]
@nicooprat Yep, the 3D is formed by straight extrusion in the z-direction with no complex roof shapes or diagonals in the z-direction possible. So all building/building:parts are rectangular!

Without any further awesomeness in Mapbox-gl-js, this is as far as we can go.

**Comment by nicooprat** [comment]
Oh too bad! Thanks for the explanation. And your work, of course :)

**Comment by klokan** [comment]
@stirringhalo did a great job here.. Big thanks belongs to him!

BTW It seems there is a space for a small improvement on the Eiffel tower shape:

<img width="930" alt="screen shot 2017-01-10 at 06 59 14" src="https://cloud.githubusercontent.com/assets/59284/21795499/9a1645d4-d702-11e6-8767-725fca5ae3dd.png">
https://www.mapbox.com/bites/00295/streets.html

**Comment by klokan** [comment]
@stirringhalo did a great job here... Big "thank you" belongs to him!

BTW It seems there is a space for a small improvement on the Eiffel tower shape:

<img width="930" alt="screen shot 2017-01-10 at 06 59 14" src="https://cloud.githubusercontent.com/assets/59284/21795499/9a1645d4-d702-11e6-8767-725fca5ae3dd.png">
https://www.mapbox.com/bites/00295/streets.html

vs

<img width="1070" alt="screen shot 2017-01-10 at 07 05 07" src="https://cloud.githubusercontent.com/assets/59284/21795587/33a47072-d703-11e6-913a-6e4e895b0556.png">

**Comment by klokan** [comment]
FYI https://github.com/mapbox/mapbox-gl-style-spec/issues/456#issuecomment-221927660

**Comment by stirringhalo** [comment]
@klokan Hmmm, I wonder if it's an issue with units/stories/feet in the base feature OSM that caused that as I've seen nothing elsewhere (especially over NY) to suggest there's a widespread issue.

I will dig into this!
